cpp/src/io/utilities/data_casting.cu

```diff
@@ -0,0 +1,925 @@
+// [c-1]!='\' &  [c]=='\' skip (already covered in skip conditions)
+// [c-1]=='\' &  [c]!=NEC && [c]!=UNICODE_SEQ, write [c]
+// [c-1]=='\' &  [c]=='u' & end_idx-curr_idx >= UNICODE_HEX_DIGIT_COUNT && hex, DECODE
+// [c+1:4]=curr_hex_val
+//        // if [c+5]=='\' & [c+6]=='u' & end_idx-curr_idx >= UNICODE_HEX_DIGIT_COUNT &&
+//        hex,DECODE [c+7:4]=next_hex_val
+//        // if [c-7]=='\' & [c-6]=='u' & end_idx-curr_idx >= UNICODE_HEX_DIGIT_COUNT &&
+//        hex,DECODE [c-5:4]=prev_hex_val prev_hex_val, curr_hex_val, next_hex_val
+//        // if prev_hex_val in high, curr_hex_val in low, skip.
+//        // if curr_hex_val in high, next_hex_val in low, write [u16]
+// if curr_hex_val not in high, write [u8]
+// before writing, find num of output characters per threads,
+// then do intra-warp/intra-block scan for out_idx
+// propagate offset from next iteration to carry forward.
+// Uses 1 warp per string or 1 block per string
+
+/**
+ * @brief Warp/Block parallel version of string_parse functor
+ *
+ * @tparam is_warp True if 1 warp per string, False if 1 block per string
+ * @tparam num_warps Number of warps per block
+ * @tparam str_tuple_it Iterator type for tuple with string pointer and its length
+ * @param str_tuples iterator of tuple with string pointer and its length
+ * @param total_out_strings Number of string rows to be processed
+ * @param str_counter Counter to keep track of processed number of strings
+ * @param null_mask Null mask
+ * @param null_count_data pointer to store null count
+ * @param options Settings for controlling string processing behavior
+ * @param d_offsets Offsets to identify where to store the results for each string
+ * @param d_chars Character array to store the characters of strings
+ */
+template <bool is_warp, size_type num_warps, typename str_tuple_it>
+__global__ void parse_fn_string_parallel(str_tuple_it str_tuples,
+                                         size_type total_out_strings,
+                                         size_type* str_counter,
+                                         bitmask_type* null_mask,
+                                         size_type* null_count_data,
+                                         cudf::io::parse_options_view const options,
+                                         size_type* d_offsets,
+                                         char* d_chars)
+{
+  constexpr auto BLOCK_SIZE =
+    is_warp ? cudf::detail::warp_size : cudf::detail::warp_size * num_warps;
+  size_type lane = is_warp ? (threadIdx.x % BLOCK_SIZE) : threadIdx.x;
+
+  // get 1-string index per warp
+  auto get_next_string = [&]() {
+    if constexpr (is_warp) {
+      size_type istring;
+      if (lane == 0) { istring = atomicAdd(str_counter, 1); }
+      __syncwarp();
+      return __shfl_sync(0xffffffff, istring, 0);
+    } else {
+      __shared__ size_type istring;
+      if (lane == 0) { istring = atomicAdd(str_counter, 1); }
+      __syncthreads();  // memory fence?
```

Comment:
question: Can I replace `__syncthreads()` with`__threadfence_block()`?
Why or why not?

Reply:
No, `__syncthreads()` is right here, because you also need to make sure that threads other than lane 0 will not progress to reading `istring ` before lane 0 has updated its value via the atomicAdd. But you can remove the comment.

Reply:
A `threadfence` only provides "happens-before" semantics for reads and writes. If you had a `threadfence` you would know that "if I observed the updated value of `istring`, then I must also be able to observe prior writes to unconnected data". Note the crucial "if". There's no synchronisation of the threads, so as @elstehle says, threads can happily walk past the `threadfence` and continue to read an old value of `istring`.

The [cppreference page](https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence) for `atomic_thread_fence` is also instructive here. Consider this example:

```c++
std::atomic<int> flag = -1;
int data;

void thread1() {
    data = 2;
    std::atomic_thread_fence(std::memory_order::seq_cst); // this is `__threadfence_block()` memory order
    flag.fetch_add(1, std::memory_order::relaxed); // this is the memory order of `atomicXXX`
}

void thread2() {
    int const v0 = flag.load(std::memory_order::relaxed);
    std::atomic_thread_fence(std::memory_order::seq_cst);
    if (v0 != -1) {
       // _If_ we see the completion of the `fetch_add`, then, because of the fence,
       // we _must_ observe the write to data.
       // _But_, without synchronisation between the threads, we are not guaranteed to observe the write
       // thread2 can execute the load _before_ thread1 runs fetch_add.
        assert(data == 2);
    }
}
```